packages/flutter/lib/src/material/chip.dart

```diff
@@ -101,83 +113,642 @@ class Chip extends StatelessWidget {
+    switch (slot) {
+      case _ChipSlot.avatar:
+        renderObject.avatar = child;
+        break;
+      case _ChipSlot.label:
+        renderObject.label = child;
+        break;
+      case _ChipSlot.deleteIcon:
+        renderObject.deleteIcon = child;
+        break;
+      case _ChipSlot.container:
+        renderObject.container = child;
+        break;
+    }
+  }
 
-    final List<Widget> children = <Widget>[];
+  @override
+  void insertChildRenderObject(RenderObject child, dynamic slotValue) {
+    assert(child is RenderBox);
+    assert(slotValue is _ChipSlot);
+    final _ChipSlot slot = slotValue;
+    _updateRenderObject(child, slot);
+    assert(renderObject.childToSlot.keys.contains(child));
+    assert(renderObject.slotToChild.keys.contains(slot));
+  }
+
+  @override
+  void removeChildRenderObject(RenderObject child) {
+    assert(child is RenderBox);
+    assert(renderObject.childToSlot.keys.contains(child));
+    _updateRenderObject(null, renderObject.childToSlot[child]);
+    assert(!renderObject.childToSlot.keys.contains(child));
+    assert(!renderObject.slotToChild.keys.contains(slot));
+  }
 
+  @override
+  void moveChildRenderObject(RenderObject child, dynamic slotValue) {
+    assert(false, 'not reachable');
+  }
+}
+
+class _ChipRenderTheme {
+  const _ChipRenderTheme({
+    @required this.avatar,
+    @required this.label,
+    @required this.deleteIcon,
+    @required this.container,
+    @required this.padding,
+    @required this.labelPadding,
+  });
+
+  final Widget avatar;
+  final Widget label;
+  final Widget deleteIcon;
+  final Widget container;
+  final EdgeInsets padding;
+  final EdgeInsets labelPadding;
+
+  @override
+  bool operator ==(dynamic other) {
+    if (identical(this, other)) {
+      return true;
+    }
+    if (other.runtimeType != runtimeType) {
+      return false;
+    }
+    final _ChipRenderTheme typedOther = other;
+    return typedOther.avatar == avatar &&
+        typedOther.label == label &&
+        typedOther.deleteIcon == deleteIcon &&
+        typedOther.container == container &&
+        typedOther.padding == padding &&
+        typedOther.labelPadding == labelPadding;
+  }
+
+  @override
+  int get hashCode {
+    return hashValues(
+      avatar,
+      label,
+      deleteIcon,
+      container,
+      padding,
+      labelPadding,
+    );
+  }
+}
+
+class _RenderChip extends RenderBox {
+  _RenderChip({
+    @required _ChipRenderTheme theme,
+    @required TextDirection textDirection,
+    this.onDeleted,
+  })  : assert(theme != null),
+        assert(textDirection != null),
+        _theme = theme,
+        _textDirection = textDirection {
+    _tap = new TapGestureRecognizer(debugOwner: this)
+      ..onTapDown = _handleTapDown
+      ..onTap = _handleTap;
+  }
+
+  // Set this to true to have outlines of the tap targets drawn over
+  // the chip.  This should never be checked in while set to 'true'.
+  static const bool _debugShowTapTargetOutlines = false;
```

Comment:
use debugPaintSizeEnabled and the debugPaint method on render objects for this kind of logic.